core/src/main/java/org/infinispan/scattered/impl/ScatteredVersionManagerImpl.java

```diff
@@ -0,0 +1,558 @@
+      }
+   }
+
+   @Override
+   public synchronized void unregisterSegment(int segment) {
+      segmentStates.set(segment, SegmentState.NOT_OWNED);
+      CompletableFuture<Void> blockedFuture = blockedFutures.get(segment);
+      if (blockedFuture != null) {
+         blockedFuture.completeExceptionally(new CacheException("The segment is no longer owned."));
+      }
+   }
+
+   @Override
+   public boolean isVersionActual(int segment, EntryVersion version) {
+      SimpleClusteredVersion clusteredVersion = (SimpleClusteredVersion) version;
+      return clusteredVersion.topologyId >= ownerTopologyIds.get(segment);
+   }
+
+   @Override
+   public void notifyKeyTransferFinished(int segment, boolean expectValues) {
+      SegmentState update = expectValues ? SegmentState.VALUE_TRANSFER : SegmentState.OWNED;
+      SegmentState previous;
+      do {
+         previous = segmentStates.get(segment);
+         // It is possible that the segment is not in KEY_TRANSFER state, but can be in BLOCKED states as well
+         // when the CONFIRM_REVOKED_SEGMENTS failed.
+         log.tracef("Finishing transfer for segment %d = %s -> %s", segment, previous, update);
+      } while (!segmentStates.compareAndSet(segment, previous, update));
+      CompletableFuture<Void> blockedFuture = blockedFutures.get(segment);
+      if (blockedFuture != null) {
+         blockedFuture.completeExceptionally(new CacheException("Segment state transition did not complete correctly."));
+      }
+      if (trace) {
+         if (expectValues) {
+            log.tracef("Node %s, segment %d has all keys in, expects value transfer", rpcManager.getAddress(), segment);
+         } else {
+            log.tracef("Node %s, segment %d did not transfer any keys, segment is owned now", rpcManager.getAddress(), segment);
+         }
+      }
+   }
+
+   @Override
+   public SegmentState getSegmentState(int segment) {
+      return segmentStates.get(segment);
+   }
+
+   @Override
+   public void setValuesTransferTopology(int topologyId) {
+      log.tracef("Node will transfer value for topology %d", topologyId);
+      synchronized (valuesLock) {
+         transferingValues = true;
+      }
+   }
+
+   @Override
+   public void notifyValueTransferFinished() {
+      for (int i = 0; i < numSegments; ++i) {
+         do {
+            SegmentState state = segmentStates.get(i);
+            switch (state) {
+               case BLOCKED:
+               case KEY_TRANSFER:
+                  blockedFutures.get(i).completeExceptionally(new CacheException("Failed to request versions"));
+                  log.warnf("Stopped applying state for segment %d in topology %d but the segment is in state %s", i, topologyId, state);
+                  // no break
+               case VALUE_TRANSFER:
+                  if (!segmentStates.compareAndSet(i, state, SegmentState.OWNED)) {
+                     continue; // goto please
```

Comment:
Mea culpa, I thought it will skip the condition. Why don't we have `goto` in Java? :,-(